In [1]:
import pandas as pd  # importamos pandas, la librería para trabajar con datos

# Cargar el dataset de razas
df = pd.read_csv('C:/Users/jezab/OneDrive/Escritorio/Adalab/Proyectos/Seidr/universos/razas_akc.csv')

# Primera exploración
print(df.shape)  # nos dice cuántas filas y columnas tiene: (filas, columnas)
df.head()        # muestra las primeras 5 filas del dataset

(277, 21)


,Unnamed: 0,description,temperament,popularity,min_height,max_height,min_weight,max_weight,min_expectancy,max_expectancy,...,grooming_frequency_value,grooming_frequency_category,shedding_value,shedding_category,energy_level_value,energy_level_category,trainability_value,trainability_category,demeanor_value,demeanor_category
0,Affenpinscher,The Affen’s apish look has been described many...,"Confident, Famously Funny, Fearless",148,22.86,29.21,3.175147,4.535924,12.0,15.0,...,0.6,2-3 Times a Week Brushing,0.6,Seasonal,0.6,Regular Exercise,0.8,Easy Training,1.0,Outgoing
1,Afghan Hound,"The Afghan Hound is an ancient breed, his whol...","Dignified, Profoundly Loyal, Aristocratic",113,63.50,68.58,22.679619,27.215542,12.0,15.0,...,0.8,Daily Brushing,0.2,Infrequent,0.8,Energetic,0.2,May be Stubborn,0.2,Aloof/Wary
2,Airedale Terrier,The Airedale Terrier is the largest of all ter...,"Friendly, Clever, Courageous",60,58.42,58.42,22.679619,31.751466,11.0,14.0,...,0.6,2-3 Times a Week Brushing,0.4,Occasional,0.6,Regular Exercise,1.0,Eager to Please,0.8,Friendly
3,Akita,"Akitas are burly, heavy-boned spitz-type dogs ...","Courageous, Dignified, Profoundly Loyal",47,60.96,71.12,31.751466,58.967008,10.0,13.0,...,0.8,Daily Brushing,0.6,Seasonal,0.8,Energetic,1.0,Eager to Please,0.6,Alert/Responsive
4,Alaskan Malamute,The Alaskan Malamute stands 23 to 25 inches at...,"Affectionate, Loyal, Playful",58,58.42,63.50,34.019428,38.555351,10.0,14.0,...,0.6,2-3 Times a Week Brushing,0.6,Seasonal,0.8,Energetic,0.4,Independent,0.8,Friendly


In [2]:
# Ver información general del dataset
df.info()  # tipos de datos de cada columna y valores nulos

<class 'pandas.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Unnamed: 0                   277 non-null    str    
 1   description                  277 non-null    str    
 2   temperament                  276 non-null    str    
 3   popularity                   198 non-null    str    
 4   min_height                   277 non-null    float64
 5   max_height                   277 non-null    float64
 6   min_weight                   275 non-null    float64
 7   max_weight                   275 non-null    float64
 8   min_expectancy               274 non-null    float64
 9   max_expectancy               274 non-null    float64
 10  group                        277 non-null    str    
 11  grooming_frequency_value     270 non-null    float64
 12  grooming_frequency_category  270 non-null    str    
 13  shedding_value               25

In [3]:
# Ver cuántos valores nulos tiene cada columna
df.isnull().sum()  # cuenta los nulos por columna — los que faltan

Unnamed: 0                      0
description                     0
temperament                     1
popularity                     79
min_height                      0
max_height                      0
min_weight                      2
max_weight                      2
min_expectancy                  3
max_expectancy                  3
group                           0
grooming_frequency_value        7
grooming_frequency_category     7
shedding_value                 20
shedding_category              20
energy_level_value              6
energy_level_category           6
trainability_value             24
trainability_category          24
demeanor_value                 25
demeanor_category              25
dtype: int64

## Análisis de valores nulos

- **popularity**: 79 nulos (28%) — no crítico, no la usaremos para asistencia
- **trainability_value**: 24 nulos (9%) — importante, necesitamos imputar o filtrar
- **demeanor_value**: 25 nulos (9%) — importante, necesitamos imputar o filtrar
- **shedding_value**: 20 nulos (7%) — secundario para el proyecto
- **energy_level_value**: 6 nulos (2%) — pocos, fácil de gestionar

### Decisión
Las columnas clave para Seiðr son:
`energy_level_value`, `trainability_value`, `demeanor_value` y `temperament`.
Los nulos en estas columnas se tratarán en el EDA del 4 bloque de trabajo.

In [4]:
# Estadísticas descriptivas de las columnas numéricas
df.describe()  # media, desviación, mínimo, máximo y percentiles de cada columna numérica

,min_height,max_height,min_weight,max_weight,min_expectancy,max_expectancy,grooming_frequency_value,shedding_value,energy_level_value,trainability_value,demeanor_value
count,277.000000,277.000000,275.000000,275.000000,274.000000,274.000000,270.000000,257.000000,271.000000,253.000000,252.000000
mean,44.225801,52.720588,17.888858,27.291416,11.306569,13.832117,0.425926,0.529183,0.712915,0.624506,0.620635
std,14.238298,15.885454,12.290600,19.061416,1.817949,2.016668,0.198306,0.189068,0.168927,0.247271,0.201713
min,12.700000,17.780000,0.000000,0.000000,0.000000,0.000000,0.200000,0.200000,0.200000,0.200000,0.200000
25%,33.020000,38.100000,8.164663,12.927383,10.000000,13.000000,0.200000,0.400000,0.600000,0.400000,0.400000
50%,45.085000,53.340000,15.875733,24.947580,12.000000,14.000000,0.400000,0.600000,0.600000,0.600000,0.600000
75%,55.880000,66.040000,22.679619,34.019428,12.000000,15.000000,0.600000,0.600000,0.800000,0.800000,0.800000
max,76.200000,88.900000,68.038855,108.862169,16.000000,19.000000,1.000000,1.000000,1.000000,1.000000,1.000000


## Observación importante — escala de valores

Las columnas numéricas clave del dataset AKC usan escala 0.2 - 1.0
mientras que Seiðr usa escala 0 - 5.

### Fórmula de conversión
`valor_seidr = round(valor_akc * 5)`

Ejemplos:
- 0.2 → 1
- 0.4 → 2
- 0.6 → 3
- 0.8 → 4
- 1.0 → 5

Esta conversión se aplicará en el bloque de limpieza de datos.

In [5]:
# Ver cuántas razas hay por grupo
df['group'].value_counts()  # cuenta razas por cada grupo AKC

group
Foundation Stock Service    68
Hound Group                 32
Sporting Group              32
Terrier Group               31
Working Group               31
Herding Group               30
Toy Group                   21
Non-Sporting Group          21
Miscellaneous Class         11
Name: count, dtype: int64

## Distribución por grupos AKC

- **Foundation Stock Service** (68): razas en proceso de reconocimiento — menos útiles para asistencia
- **Hound y Sporting** (32 cada uno): buena representación
- **Terrier, Working y Herding** (~30 cada uno): grupos con razas de asistencia conocidas
- **Toy y Non-Sporting** (21 cada uno): razas pequeñas, relevantes para perfiles con necesidades específicas

### Decisión
Filtrar o marcar las razas de Foundation Stock Service como prioridad baja
para el proyecto, ya que tienen menos datos de comportamiento documentados.

In [6]:
# Razas con mayor capacidad de entrenamiento — candidatas a asistencia
df[['Unnamed: 0', 'temperament', 'energy_level_value', 'trainability_value', 'demeanor_value']] \
    .dropna() \
    .sort_values('trainability_value', ascending=False) \
    .head(20)

,Unnamed: 0,temperament,energy_level_value,trainability_value,demeanor_value
2,Airedale Terrier,"Friendly, Clever, Courageous",0.6,1.0,0.8
7,American Eskimo Dog,"Playful, Perky, Smart",0.8,1.0,1.0
3,Akita,"Courageous, Dignified, Profoundly Loyal",0.8,1.0,0.6
12,American Water Spaniel,"Eager, Happy, Charming",0.6,1.0,0.8
10,American Leopard Hound,"Sociable, Energetic, Intelligent",0.8,1.0,0.6
17,Australian Shepherd,"Smart, Work-Oriented, Exuberant",0.8,1.0,0.6
16,Australian Kelpie,"Loyal, Alert, Intelligent",0.8,1.0,0.6
54,Braque du Bourbonnais,"Affectionate, Adaptable, Gentle",0.6,1.0,0.6
38,Biewer Terrier,"Intelligent, Devoted, Amusing",0.4,1.0,1.0
57,Brittany,"Bright, Fun-Loving, Upbeat",1.0,1.0,0.8


## Razas candidatas para asistencia — primeras conclusiones

Las razas con trainability_value = 1.0 son fácilmente entrenables
pero eso no las convierte automáticamente en las mejores para asistencia.

### Criterios reales para un perro de asistencia
Un buen perro de asistencia no es el más entrenable.
Es el que mejor encaja con el perfil específico de su usuario.

**Los criterios clave son:**
- Vínculo emocional con el usuario (demeanor_value)
- Temperamento afectuoso y familiar (temperament)
- Adaptabilidad al entorno del usuario
- Energía compatible con el estilo de vida del usuario
- Tamaño adecuado para el contexto

**Ejemplos reales que lo confirman:**
- Spitz japonés: trainability moderada, vínculo excepcional, afectuoso
- Groenendael: inteligente, protectora, vínculo muy intenso con su persona

### Decisión para Seiðr
El algoritmo NO buscará la raza más entrenable.
Buscará la raza cuyo perfil de temperamento, energía y vínculo
encaje mejor con el perfil de 8 dimensiones del usuario.

In [7]:
# Buscar razas específicas mencionadas como perros de asistencia con experiencia personal
razas_buscar = ['Spitz', 'Groenendael', 'Belgian', 'Malinois']

# Filtrar razas que contengan esos nombres
df[df['Unnamed: 0'].str.contains('|'.join(razas_buscar), case=False, na=False)][
    ['Unnamed: 0', 'temperament', 'energy_level_value', 'trainability_value', 'demeanor_value']
]

,Unnamed: 0,temperament,energy_level_value,trainability_value,demeanor_value
30,Belgian Laekenois,"Affectionate, Alert, Intelligent",0.8,0.8,0.4
31,Belgian Malinois,"Confident, Smart, Hardworking",0.8,0.8,0.4
32,Belgian Sheepdog,"Bright, Watchful, Serious-Minded",0.8,0.8,0.4
33,Belgian Tervuren,"Courageous, Alert, Intelligent",0.8,0.8,0.4
108,Finnish Spitz,"Friendly, Good-Natured, Lively",1.0,0.6,0.8
116,German Spitz,"Devoted, Lively, Attentive",0.6,1.0,0.4
143,Japanese Spitz,"Loyal, Playful, Smart",0.6,0.8,0.8


## Limitación del dataset AKC

El campo `demeanor_value` mide sociabilidad general con extraños,
no el vínculo con el usuario específico.

Un Groenendael tiene demeanor 0.4 (reservado con extraños)
pero vínculo excepcional con su persona de referencia.

### Decisión
Necesitamos añadir un campo `vinculo_usuario` en razas_asistencia.csv
que capture esta dimensión que el dataset AKC no mide.
Este campo se rellenará con investigación adicional de fuentes especializadas.

In [8]:
# Renombrar columna sin nombre y limpiar nombres de columnas
df = df.rename(columns={'Unnamed: 0': 'breed'})  # renombramos a breed (raza en inglés)

# Verificar que el cambio se hizo bien
df.columns.tolist()  # muestra todos los nombres de columnas

['breed',
 'description',
 'temperament',
 'popularity',
 'min_height',
 'max_height',
 'min_weight',
 'max_weight',
 'min_expectancy',
 'max_expectancy',
 'group',
 'grooming_frequency_value',
 'grooming_frequency_category',
 'shedding_value',
 'shedding_category',
 'energy_level_value',
 'energy_level_category',
 'trainability_value',
 'trainability_category',
 'demeanor_value',
 'demeanor_category']

## CORREGIMOS COSAS COMO RENOMBRAR COLUMNA 0 POR BREED O LA ESCALA DE PUNTUACIÓN

In [9]:
# Convertir escala AKC (0.2-1.0) a escala Seiðr (0-5)
columnas_escala = [
    'grooming_frequency_value',
    'shedding_value', 
    'energy_level_value',
    'trainability_value',
    'demeanor_value'
]

for col in columnas_escala:  # recorremos cada columna numérica clave
    df[col + '_seidr'] = (df[col] * 5).round().astype('Int64')  # multiplicamos por 5 y redondeamos

# Verificar la conversión con un ejemplo
df[['breed', 'energy_level_value', 'energy_level_value_seidr', 
    'trainability_value', 'trainability_value_seidr']].head(10)

,breed,energy_level_value,energy_level_value_seidr,trainability_value,trainability_value_seidr
0,Affenpinscher,0.6,3,0.8,4
1,Afghan Hound,0.8,4,0.2,1
2,Airedale Terrier,0.6,3,1.0,5
3,Akita,0.8,4,1.0,5
4,Alaskan Malamute,0.8,4,0.4,2
5,American Bulldog,0.8,4,0.6,3
6,American English Coonhound,0.8,4,0.6,3
7,American Eskimo Dog,0.8,4,1.0,5
8,American Foxhound,0.8,4,0.4,2
9,American Hairless Terrier,NaN,<NA>,NaN,<NA>


## Conversión de escala completada

Las columnas numéricas clave ahora tienen su equivalente en escala Seiðr (0-5):
- energy_level_value_seidr
- trainability_value_seidr
- demeanor_value_seidr
- grooming_frequency_value_seidr
- shedding_value_seidr

Los valores nulos se conservan como NA — se tratarán en el bloque de limpieza.

### EDA 

In [10]:
# ============================================================
# EDA COMPLETO — Limpieza de datos
# ============================================================

# 1. Ver nulos actuales
print("=== NULOS POR COLUMNA ===")
print(df.isnull().sum())
print(f"\nTotal filas: {len(df)}")
print(f"Total columnas: {len(df.columns)}")

=== NULOS POR COLUMNA ===
breed                              0
description                        0
temperament                        1
popularity                        79
min_height                         0
max_height                         0
min_weight                         2
max_weight                         2
min_expectancy                     3
max_expectancy                     3
group                              0
grooming_frequency_value           7
grooming_frequency_category        7
shedding_value                    20
shedding_category                 20
energy_level_value                 6
energy_level_category              6
trainability_value                24
trainability_category             24
demeanor_value                    25
demeanor_category                 25
grooming_frequency_value_seidr     7
shedding_value_seidr              20
energy_level_value_seidr           6
trainability_value_seidr          24
demeanor_value_seidr              25
dtype: int64

In [12]:
# ============================================================
# EDA COMPLETO — Limpieza de datos
# ============================================================

# 1. Ver nulos actuales — cuántos valores faltan por columna
print("=== NULOS POR COLUMNA ===")
print(df.isnull().sum())  # suma los nulos de cada columna
print(f"\nTotal filas: {len(df)}")     # número de razas
print(f"Total columnas: {len(df.columns)}")  # número de características

# 2. Comprobar si hay filas duplicadas
print("\n=== LIMPIEZA ===")
print(f"Duplicados: {df.duplicated().sum()}")  # 0 = no hay duplicados, perfecto

# 3. Rellenar nulos en columnas clave con la mediana
# Usamos mediana (valor del medio) en lugar de media (promedio)
# porque la mediana no se ve afectada por valores extremos
columnas_clave = [
    'energy_level_value',       # nivel de energía en escala AKC (0.2-1.0)
    'trainability_value',       # capacidad de entrenamiento en escala AKC
    'demeanor_value',           # comportamiento/temperamento en escala AKC
    'energy_level_value_seidr', # nivel de energía convertido a escala Seiðr (0-5)
    'trainability_value_seidr', # entrenabilidad en escala Seiðr
    'demeanor_value_seidr'      # comportamiento en escala Seiðr
]

for col in columnas_clave:  # recorremos cada columna clave
    mediana = df[col].median()           # calculamos la mediana de esa columna
    nulos_antes = df[col].isnull().sum() # guardamos cuántos nulos había
    df[col] = df[col].fillna(mediana)    # rellenamos los nulos con la mediana
    print(f"{col}: {nulos_antes} nulos → rellenados con mediana ({mediana})")

# 4. Verificar que no quedan nulos en columnas clave
print(f"\nNulos restantes en columnas clave:")
print(df[columnas_clave].isnull().sum())  # debería ser 0 en todas

=== NULOS POR COLUMNA ===
breed                              0
description                        0
temperament                        1
popularity                        79
min_height                         0
max_height                         0
min_weight                         2
max_weight                         2
min_expectancy                     3
max_expectancy                     3
group                              0
grooming_frequency_value           7
grooming_frequency_category        7
shedding_value                    20
shedding_category                 20
energy_level_value                 0
energy_level_category              6
trainability_value                 0
trainability_category             24
demeanor_value                     0
demeanor_category                 25
grooming_frequency_value_seidr     7
shedding_value_seidr              20
energy_level_value_seidr           0
trainability_value_seidr           0
demeanor_value_seidr               0
dtype: int64

In [13]:
import plotly.express as px  # librería para gráficos interactivos

# ============================================================
# VISUALIZACIÓN 1 — Distribución de razas por grupo AKC
# ============================================================

# Contamos cuántas razas hay en cada grupo
conteo_grupos = df['group'].value_counts().reset_index()  # reset_index convierte el resultado en DataFrame
conteo_grupos.columns = ['grupo', 'cantidad']  # renombramos las columnas

# Creamos el gráfico de barras
fig = px.bar(
    conteo_grupos,                    # datos a usar
    x='grupo',                        # eje X: nombre del grupo
    y='cantidad',                     # eje Y: número de razas
    title='Distribución de razas por grupo AKC',  # título del gráfico
    color='cantidad',                 # colorear barras según la cantidad
    color_continuous_scale='Viridis', # paleta de colores
    text='cantidad'                   # mostrar el número encima de cada barra
)

fig.update_layout(
    xaxis_tickangle=-45,  # rotar etiquetas del eje X para que no se solapen
    showlegend=False       # ocultar leyenda — no es necesaria aquí
)

fig.show()  # mostrar el gráfico

In [14]:
# ============================================================
# VISUALIZACIÓN 2 — Distribución de nivel de energía
# ============================================================

# Histograma: cuántas razas tienen cada nivel de energía en escala Seiðr
fig2 = px.histogram(
    df,                              # datos a usar
    x='energy_level_value_seidr',   # eje X: nivel de energía en escala Seiðr (0-5)
    title='Distribución del nivel de energía en razas AKC (escala Seiðr 0-5)',
    nbins=5,                         # 5 barras — una por cada valor posible (1,2,3,4,5)
    color_discrete_sequence=['#7F77DD'],  # color morado de Seiðr
    text_auto=True                   # mostrar cantidad encima de cada barra
)

fig2.update_layout(
    xaxis_title='Nivel de energía (0=muy tranquila, 5=muy activa)',
    yaxis_title='Número de razas',
    bargap=0.1  # pequeño espacio entre barras
)

fig2.show()

## Conclusión — Distribución de energía

La mayoría de razas AKC tienen nivel de energía 3-4 (moderado-alto).
Las razas con energía 1-2 (muy tranquilas) son escasas: solo 20 razas.
Esto es relevante para perfiles con SPD, TEA o hipersensibilidad sensorial
que necesitan un compañero de baja estimulación.

In [15]:
# ============================================================
# VISUALIZACIÓN 3 — Comparativa entrenabilidad vs comportamiento
# ============================================================

# Scatter plot: relación entre entrenabilidad y comportamiento
# Cada punto es una raza — buscamos las que tienen ambos valores altos
fig3 = px.scatter(
    df,                                    # datos
    x='trainability_value_seidr',          # eje X: entrenabilidad (0-5)
    y='demeanor_value_seidr',              # eje Y: comportamiento/vínculo (0-5)
    hover_name='breed',                    # al pasar el ratón muestra el nombre de la raza
    color='energy_level_value_seidr',      # color según nivel de energía
    color_continuous_scale='Viridis',      # paleta de colores
    title='Entrenabilidad vs Comportamiento — Candidatas a asistencia',
    size_max=10                            # tamaño máximo de los puntos
)

fig3.update_layout(
    xaxis_title='Entrenabilidad (0=difícil, 5=muy fácil)',
    yaxis_title='Comportamiento/Vínculo (0=reservado, 5=muy sociable)',
)

# Añadir líneas de referencia para identificar el cuadrante ideal
fig3.add_hline(y=4, line_dash='dash', line_color='red',   # línea horizontal en y=4
               annotation_text='Vínculo alto')
fig3.add_vline(x=4, line_dash='dash', line_color='red',   # línea vertical en x=4
               annotation_text='Entrenabilidad alta')

fig3.show()  # mostrar gráfico

## Conclusión — Entrenabilidad vs Comportamiento

El cuadrante ideal (entrenabilidad ≥4 Y vínculo ≥4) tiene pocas razas.
Las destacadas son:
- **Welsh Springer Spaniel**: perfil equilibrado — entrenabilidad 4, vínculo 4
- **Havanese**: vínculo excepcional (5), energía baja — ideal para perfiles SPD o TEA
- **Papillon**: vínculo muy alto, entrenabilidad moderada — apoyo emocional

Confirmación del análisis anterior: el criterio de selección para Seiðr
NO es máxima entrenabilidad sino equilibrio entre vínculo y entrenabilidad
adaptado al perfil del usuario.

In [16]:
# ============================================================
# VISUALIZACIÓN 4 — Top 15 razas candidatas para asistencia
# ============================================================

# Creamos una puntuación compuesta para asistencia
# Combinamos entrenabilidad (40%) + vínculo (40%) + energía moderada (20%)
# La energía moderada se calcula como: 5 - abs(energy - 3) 
# Es decir, penalizamos las razas con energía muy alta o muy baja
# porque para asistencia lo ideal es energía media

df['puntuacion_asistencia'] = (
    df['trainability_value_seidr'] * 0.4 +   # 40% entrenabilidad
    df['demeanor_value_seidr'] * 0.4 +        # 40% vínculo/comportamiento
    (5 - abs(df['energy_level_value_seidr'] - 3)) * 0.2  # 20% energía moderada
)

# Ordenamos por puntuación y cogemos las 15 mejores
top15 = (df[['breed', 'puntuacion_asistencia', 
              'trainability_value_seidr', 
              'demeanor_value_seidr',
              'energy_level_value_seidr']]
         .sort_values('puntuacion_asistencia', ascending=False)  # orden descendente
         .head(15))  # las 15 primeras

# Gráfico de barras horizontal
fig4 = px.bar(
    top15,
    x='puntuacion_asistencia',   # eje X: puntuación total
    y='breed',                    # eje Y: nombre de la raza
    orientation='h',              # barras horizontales
    color='puntuacion_asistencia', # color según puntuación
    color_continuous_scale='Viridis',
    title='Top 15 razas candidatas para asistencia en Seiðr',
    text=top15['puntuacion_asistencia'].round(2)  # mostrar puntuación en la barra
)

fig4.update_layout(
    yaxis={'categoryorder': 'total ascending'},  # ordenar de menor a mayor visualmente
    xaxis_title='Puntuación de asistencia (0-5)',
    yaxis_title='',
    showlegend=False
)

fig4.show()

# Mostrar tabla con los datos completos
print("\nTop 15 razas candidatas:")
print(top15.to_string(index=False))


Top 15 razas candidatas:
                   breed  puntuacion_asistencia  trainability_value_seidr  demeanor_value_seidr  energy_level_value_seidr
                Papillon                    5.0                         5                     5                         3
          Boston Terrier                    5.0                         5                     5                         3
          Cocker Spaniel                    5.0                         5                     5                         3
     American Eskimo Dog                    4.8                         5                     5                         4
          Biewer Terrier                    4.8                         5                     5                         2
            Poodle (Toy)                    4.6                         5                     4                         3
              Leonberger                    4.6                         5                     4                         

## Conclusión — Top 15 razas candidatas

El algoritmo de puntuación equilibra entrenabilidad (40%) + vínculo (40%) 
+ energía moderada (20%). El resultado es coherente con la realidad:
el Labrador Retriever aparece en el top, confirmando que el criterio funciona.

**Decisión de diseño importante:**
El tamaño NO es un factor penalizador en Seiðr.
Un perro de asistencia no tiene por qué ser grande.
Depende exclusivamente de las necesidades del usuario:
- Perro de alerta médica o apoyo emocional → tamaño irrelevante
- Perro guía → necesita tamaño adecuado
- Perro de servicio físico → necesita fuerza proporcional a la tarea

Seiðr filtrará por perfil de usuario, no por tamaño estándar.

In [17]:
# ============================================================
# VISUALIZACIÓN 5 — Organizaciones de asistencia por CCAA
# ============================================================

import pandas as pd  # ya importado pero lo dejamos para claridad

# Cargar el CSV de organizaciones que scrapeamos
df_orgs = pd.read_csv('../universos/recursos_reales/organizaciones_espana.csv')  # cargamos el CSV

# Contar organizaciones por comunidad autónoma
conteo_ccaa = df_orgs['ccaa'].value_counts().reset_index()  # contamos por CCAA
conteo_ccaa.columns = ['ccaa', 'num_organizaciones']        # renombramos columnas

# Gráfico de barras horizontal
fig5 = px.bar(
    conteo_ccaa,
    x='num_organizaciones',          # eje X: número de organizaciones
    y='ccaa',                        # eje Y: comunidad autónoma
    orientation='h',                 # barras horizontales
    color='num_organizaciones',      # color según cantidad
    color_continuous_scale='Viridis',
    title='Organizaciones de perros de asistencia por Comunidad Autónoma',
    text='num_organizaciones'        # mostrar número en cada barra
)

fig5.update_layout(
    yaxis={'categoryorder': 'total ascending'},  # ordenar de menor a mayor
    xaxis_title='Número de organizaciones',
    yaxis_title='',
    showlegend=False
)

fig5.show()

# Mostrar también qué tipos de asistencia hay disponibles
print("\n=== TIPOS DE ASISTENCIA DISPONIBLES ===")
tipos = df_orgs['tipo_asistencia'].str.split(';').explode()  # separar tipos por ;
print(tipos.value_counts())  # contar cuántas organizaciones ofrecen cada tipo


=== TIPOS DE ASISTENCIA DISPONIBLES ===
tipo_asistencia
perro_servicio    30
perro_tea         30
perro_alerta      28
perro_señal       26
perro_guia        21
Name: count, dtype: int64


## Conclusión — Organizaciones por CCAA y tipos de asistencia

**Distribución geográfica:**
Madrid concentra el 32% de las organizaciones (12/38).
Canarias tiene solo 1 organización acreditada (Terapican).
Hay CCAA sin ninguna organización en este directorio — limitación del dataset.

**Tipos de asistencia disponibles:**
- Perro de servicio y perro TEA: 30 organizaciones cada uno — los más ofertados
- Perro de alerta médica: 28 — muy bien representado
- Perro señal (discapacidad auditiva): 26
- Perro guía: 21 — el más conocido pero no el más ofertado

**Implicación para Seiðr:**
El módulo 2 filtrará organizaciones por CCAA del usuario Y por tipo de asistencia
según su perfil. Un usuario de Canarias verá Terapican con su información completa.
Un usuario de Madrid tendrá 12 opciones filtradas por su necesidad específica.

In [21]:
fig6 = px.imshow(
    correlacion,
    labels=dict(color="Correlación"),
    x=nombres,
    y=nombres,
    color_continuous_scale='RdBu',
    zmin=-1, zmax=1,
    title='Correlación entre variables clave del dataset AKC',
    text_auto=True
)

fig6.show()

## Conclusión — Mapa de correlaciones

**Hallazgo clave:** Energía y Vínculo tienen correlación 0.
Son dimensiones completamente independientes.
Hay razas energéticas con vínculo alto Y razas tranquilas con vínculo alto.
Esto confirma que el algoritmo de Seiðr puede encontrar el perfil exacto
para cada usuario sin que la energía limite el vínculo ni viceversa.

**Nota técnica:** La alta correlación entre Entrenabilidad/Vínculo y
Punt.Asistencia es esperada — son los factores que más peso tienen en la fórmula.
Si se decide dar más peso al vínculo que a la entrenabilidad,
solo hay que ajustar los coeficientes de la fórmula.

In [22]:
# ============================================================
# VISUALIZACIÓN 7 — Perfil medio de asistencia por grupo AKC
# ============================================================

# Calculamos la media de las variables clave por grupo
# Esto nos dice qué grupos tienen mejor perfil general para asistencia
perfil_grupos = df.groupby('group')[[ 
    'energy_level_value_seidr',       # energía media del grupo
    'trainability_value_seidr',       # entrenabilidad media
    'demeanor_value_seidr',           # vínculo medio
    'puntuacion_asistencia'           # puntuación media de asistencia
]].mean().round(2).reset_index()      # calculamos media y redondeamos

# Renombramos para el gráfico
perfil_grupos.columns = ['Grupo', 'Energía', 'Entrenabilidad', 'Vínculo', 'Punt.Asistencia']

# Ordenamos por puntuación de asistencia
perfil_grupos = perfil_grupos.sort_values('Punt.Asistencia', ascending=False)

# Gráfico de barras agrupadas — comparamos las 3 dimensiones por grupo
fig7 = px.bar(
    perfil_grupos,
    x='Grupo',                        # eje X: grupo AKC
    y=['Entrenabilidad', 'Vínculo', 'Energía'],  # tres barras por grupo
    title='Perfil medio de asistencia por grupo AKC',
    barmode='group',                  # barras agrupadas, no apiladas
    color_discrete_map={              # colores específicos para cada variable
        'Entrenabilidad': '#7F77DD',  # morado Seiðr
        'Vínculo': '#1D9E75',         # verde
        'Energía': '#BA7517'          # ámbar
    }
)

fig7.update_layout(
    xaxis_tickangle=-45,              # rotar etiquetas para que no se solapen
    yaxis_title='Puntuación media (0-5)',
    xaxis_title='',
    legend_title='Dimensión'
)

fig7.show()

# Mostrar tabla resumen ordenada por puntuación
print("\nRanking de grupos por puntuación de asistencia:")
print(perfil_grupos[['Grupo', 'Entrenabilidad', 'Vínculo', 'Energía', 'Punt.Asistencia']].to_string(index=False))


Ranking de grupos por puntuación de asistencia:
                   Grupo  Entrenabilidad  Vínculo  Energía  Punt.Asistencia
          Sporting Group             4.0     3.34     4.06             3.72
               Toy Group             3.1     3.57     2.95             3.64
     Miscellaneous Class            3.82     3.09     3.91             3.55
      Non-Sporting Group             3.0     3.38     3.29             3.46
           Terrier Group            2.84     3.39     3.42             3.41
           Herding Group            3.53     2.73     3.83             3.33
           Working Group             3.1      3.0     3.71             3.28
Foundation Stock Service            2.94     2.75     3.38             3.13
             Hound Group            2.31     3.22     3.56             3.08


## Conclusión — Perfil por grupos AKC

**Sporting Group** — mejor perfil general para asistencia:
entrenabilidad alta (4.0) con energía moderada-alta.
Spaniels, Retrievers y Setters pertenecen a este grupo.

**Toy Group** — mejor perfil para asistencia emocional:
vínculo alto (3.6) con energía moderada-baja (3.0).
Ideal para perfiles TEA, SPD o que necesitan calma.

**Hound Group** — menos adecuado para asistencia:
entrenabilidad más baja (2.3). Independientes por naturaleza.

**Implicación para Seiðr:**
El grupo AKC puede usarse como filtro secundario en el algoritmo
para refinar la selección de raza según el perfil del usuario.

In [23]:
# ============================================================
# GUARDAR DATASET LIMPIO
# ============================================================

# Guardamos el dataframe limpio con las columnas en escala Seiðr
# Este es el archivo que usará el algoritmo de Seiðr
df.to_csv('../universos/razas_akc_limpio.csv', 
          index=False,           # no guardamos el índice como columna
          encoding='utf-8-sig')  # encoding para acentos y caracteres especiales

print(f"Dataset limpio guardado: {len(df)} razas, {len(df.columns)} columnas")
print(f"\nColumnas disponibles:")
print(df.columns.tolist())

Dataset limpio guardado: 277 razas, 27 columnas

Columnas disponibles:
['breed', 'description', 'temperament', 'popularity', 'min_height', 'max_height', 'min_weight', 'max_weight', 'min_expectancy', 'max_expectancy', 'group', 'grooming_frequency_value', 'grooming_frequency_category', 'shedding_value', 'shedding_category', 'energy_level_value', 'energy_level_category', 'trainability_value', 'trainability_category', 'demeanor_value', 'demeanor_category', 'grooming_frequency_value_seidr', 'shedding_value_seidr', 'energy_level_value_seidr', 'trainability_value_seidr', 'demeanor_value_seidr', 'puntuacion_asistencia']


## Resumen del EDA completo

**Dataset:** 277 razas AKC · 27 columnas · 0 duplicados

**Limpieza aplicada:**
- Nulos en columnas clave rellenados con la mediana
- Columna breed renombrada desde Unnamed: 0
- Escala AKC (0.2-1.0) convertida a escala Seiðr (0-5)

**Variables creadas:**
- energy_level_value_seidr, trainability_value_seidr, demeanor_value_seidr
- puntuacion_asistencia: fórmula propia (entrenabilidad 40% + vínculo 40% + energía moderada 20%)

**Hallazgos principales:**
1. Energía y vínculo son dimensiones independientes (correlación 0)
2. Sporting Group tiene el mejor perfil general para asistencia
3. Toy G